src/util/fetch.js

```diff
@@ -0,0 +1,196 @@
+ * });
+ *
+ * request.on('error', e => {
+ *     console.warn('fetchFile error: ', e);
+ * });
+ */
+export default function fetchFile(options) {
+    if (!options) {
+        throw new Error('fetch options missing');
+    }
+    if (!options.url) {
+        throw new Error('fetch url missing');
+    }
+    const instance = new Observer();
+    let fetchHeaders = new Headers();
+    let fetchRequest = new Request(options.url);
+
+    // add ability to abort
+    instance.controller = new AbortController();
+
+    // check if headers have to be added
+    if (options && options.requestHeaders) {
+        // add custom request headers
+        options.requestHeaders.forEach(header => {
+            fetchHeaders.append(header.key, header.value);
+        });
+    }
+
+    // parse fetch options
+    const responseType = options.responseType || 'json';
+    const fetchOptions = {
+        method: options.method || 'GET',
+        headers: fetchHeaders,
+        mode: options.mode || 'cors',
+        credentials: options.credentials || 'same-origin',
+        cache: options.cache || 'default',
+        redirect: options.redirect || 'follow',
+        referrer: options.referrer || 'client',
+        signal: instance.controller.signal
+    };
+
+    fetch(fetchRequest, fetchOptions)
+        .then(response => {
+            // store response reference
+            instance.response = response;
+
+            let progressAvailable = true;
+            if (!response.body) {
+                // ReadableStream is not yet supported in this browser
+                // see https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream
+                progressAvailable = false;
+            }
+
+            // Server must send CORS header "Access-Control-Expose-Headers: content-length"
+            const contentLength = response.headers.get('content-length');
+            if (contentLength === null) {
+                // Content-Length server response header missing.
+                // Don't evaluate download progress if we can't compare against a total size
+                // see https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Access-Control-Expose-Headers
+                progressAvailable = false;
+            }
+
+            // not able to check download progress
+            if (!progressAvailable) {
+                return response;
+            }
+
+            const total = parseInt(contentLength, 10);
+            instance._reader = response.body.getReader();
+            let loaded = 0;
+
+            instance.onProgress = e => {
+                instance.fireEvent('progress', e);
+            };
+
+            return new Response(
+                new ReadableStream({
+                    start(controller) {
+                        read();
+
+                        /**
+                         * start reading
+                         * @private
+                         */
+                        function read() {
```

Comment:
Could this function be defined somewhere outside to reduce the nestedness?

Reply:
sounds good, I'll probably put it in a separate class or something.